### ISSUE TITLE
example_deconv_train crashes in the end

### ISSUE THREAD

**Issue by reunanen** [nyanp/tiny-cnn#199]
The [deconvolution example](https://github.com/nyanp/tiny-cnn/tree/master/examples/deconv) pretty much consistently crashes in the end, and sometimes even before the end (but this is random, as is usual with memory access problems). Attaching some debugger screenshots.

![image](https://cloud.githubusercontent.com/assets/2297572/16227236/27d79dc2-37b8-11e6-935b-125aeb7db1f3.png)

![image](https://cloud.githubusercontent.com/assets/2297572/16227278/5ecca7be-37b8-11e6-9ce3-405d9585a2f7.png)

I'm using Visual Studio 2015.

Can anyone reproduce this?

**Comment by reunanen** [comment]
It looks like the `ppi` [here](https://github.com/nyanp/tiny-cnn/blob/705fd656d67fd7f32254f432161b354f7905061b/tiny_cnn/layers/deconvolutional_layer.h#L215) points outside the memory allocated for vector `in`.

@Wangyida, any comments?

**Comment by reunanen** [comment]
It also looks like there's a similar issue [here](https://github.com/nyanp/tiny-cnn/blob/705fd656d67fd7f32254f432161b354f7905061b/tiny_cnn/layers/deconvolutional_layer.h#L281).

**Comment by Wangyida** [comment]
@reunanen Have you pull the latest version from branch feat/generic-computational-graph? Deconvolution codes are updated many times these days and it works on Mac OS and Ubuntu 14.04. I suggest you running the test codes for deconvolution first.

**Comment by reunanen** [comment]
@Wangyida No, I'm on the master branch. Will check feat/generic-computational-graph and let you know.

**Comment by reunanen** [comment]
@Wangyida In branch _feat/generic-computational-graph_, I have no issues with `deLaNet`. However, `DeAE` still makes my debugger complain a lot. But that may be something different, and I guess the branch is under development anyway.

If you don't mind, I'd leave the issue open until the master branch has a working example of how deconvolution can be used.

**Comment by Wangyida** [comment]
@reunanen  As for the DeAE, I think there might be some memory bugs in the unpooling layer, this structure is still under developing. You can have a try on the visual example. Using deconvolution alone will works at now. You can have a look on the readme.md and try on this example

**Comment by edgarriba** [comment]
/cc @nyanp 

We need to synchronize both branches.
IMO, we could it once the first backend version is working with NNPACK. Besides, batch processing should be adapted to new API.
